src/docker.c

```diff
@@ -0,0 +1,1137 @@
+  dispatch_container_cpu_stats(resource->stats->cpu_stats, hostname);
+  dispatch_container_memory_stats(resource->stats->memory_stats, hostname);
+  dispatch_container_network_stats(resource->stats->network_stats, hostname);
+  sfree(hostname);
+}
+
+static int dispatch_stats_all(void) {
+  char **container_list = NULL;
+  int count = get_container_list(&container_list, DOCKER_SOCKET,
+				 DOCKER_VERSION);
+  if (count == 0) {
+    DEBUG("docker: No containers running on this machine.");
+    goto leave;
+  } else if (count == -1) {
+    ERROR("docker: Unable to parse container information.");
+    goto leave;
+  }
+  container_resource_t **containers = (container_resource_t **)
+      calloc(count, sizeof(container_resource_t *));
+  for (int i = 0; i < count; i++) {
+    containers[i] = (container_resource_t *)
+        calloc(1, sizeof(container_resource_t));
+    containers[i]->id = container_list[i];
+    containers[i]->stats = get_stats_for_container(containers[i]->id,
+        DOCKER_SOCKET, DOCKER_VERSION);
+  }
+  for (int i = 0; i < count; i++) {
+    if (containers[i]->stats != NULL) {
+      dispatch_container_stats(containers[i]);
+    }
+  }
+  for (int i = 0; i < count; i++) {
+    if (containers[i]->stats != NULL) {
+      free_stats(containers[i]->stats);
+    }
+    sfree(containers[i]);
+  }
+  sfree(containers);
+ leave:
+  free_list((void ***) &container_list, count);
+  return 0;
+}
+
+//==============================================================================
+//==============================================================================
+//==============================================================================
+// Collectd plugin load and configuration fucntionality.
+//==============================================================================
+//==============================================================================
+//==============================================================================
+
+static int docker_config(const char *key, const char *value) {
+  if (strcmp(key, "Socket") == 0) {
+    config_socket = (const char *) sstrdup(value);
+    return 0;
+  }
+
+  if (strcmp(key, "Version") == 0) {
+    config_version = (const char *) sstrdup(value);
+    return 0;
+  }
+  WARNING("docker: Unknown config option found. Key: %s, Value: %s", key,
+	  value);
+  return -1;
+}
+
+static int docker_init(void) {
+  if (DOCKER_VERSION == NULL) {
+    DOCKER_VERSION = (const char *)
+	(config_version != NULL ? config_version : DEFAULT_VERSION);
+  }
+
+  if (DOCKER_SOCKET == NULL) {
+    DOCKER_SOCKET = (const char *)
+	(config_socket != NULL ? config_socket : DEFAULT_SOCKET);
+  }
+  cpu_hist_values = c_avl_create((int(*)(const void *, const void *))&strcmp);
+  if (cpu_hist_values == NULL) {
+    ERROR("docker: c_avl_create failed.");
+    return -1;
+  }
+  return 0;
+}
+
+static int docker_shutdown(void) {
+  while (1) {
```

Comment:
How about:

void *key;
void* value;
while (c_avl_pick(cpu_hist_values, &key, &value) == 0) {
    <what you have in the 'if' clause>
}
c_avl_destroy(cpu_hist_values);
return 0;